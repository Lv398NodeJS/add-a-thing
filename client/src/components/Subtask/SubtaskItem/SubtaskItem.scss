$subtask-bg-onhover: linear-gradient(to right, #ececec, #f6f6f6);
$delete-button: #f6f6f6;

.subtask-row:hover {
  background: $subtask-bg-onhover;
}

.subtask-row:hover .subtask-dropdown-toggle {
  transform: scale(1);
  transition: transform 150ms ease;
}

.subtask-dropdown-toggle {
  &,
  &:hover,
  &:link,
  &:focus,
  &:active {
    background: inherit;
    transform: scale(0);
    border: none;
    outline: none;
  }
}
.subtask-dropdown-toggle:after {
  display: none;
}
.threedots-icon {
  width: 0.9rem;
}

// Styling for checked/unchecked checkbox and its label
.custom-checkbox {
  .custom-control-input:checked {
    ~ .custom-control-label::before {
      background-color: var(--green);
      border-color: var(--green);
    }
    ~ .custom-control-label {
      color: var(--green);
      text-decoration: line-through;
    }
  }
  .custom-control-input:checked:disabled {
    ~ .custom-control-label::before {
      background-color: var(--gray);
      border-color: var(--gray);
    }
    ~ .custom-control-label {
      color: var(--gray);
    }
  }
}

.subtask-save-button {
  img {
    width: 15px;
  }
} 