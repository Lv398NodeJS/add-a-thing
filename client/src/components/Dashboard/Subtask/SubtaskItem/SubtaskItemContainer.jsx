import React, { Component } from 'react';
import SubtaskItem from './SubtaskItem';
import SubtaskEdit from './SubtaskEdit';

export class SubtaskItemContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditMode: false,
    };
  }

  toggleEditMode = () => {
    this.setState(prevState => ({
      isEditMode: !prevState.isEditMode,
    }));
  }

  render() {
    const {
      _id,
      name,
      completed,
      taskStatus,
    } = this.props;
    const { isEditMode } = this.state;
    return (
      isEditMode
        ? (
          <SubtaskEdit
            _id={_id}
            name={name}
            toggleEditMode={this.toggleEditMode}
          />
        )
        : (
          <SubtaskItem
            _id={_id}
            name={name}
            completed={completed}
            taskStatus={taskStatus}
            toggleEditMode={this.toggleEditMode}
          />
        )
    );
  }
}

export default SubtaskItemContainer;
