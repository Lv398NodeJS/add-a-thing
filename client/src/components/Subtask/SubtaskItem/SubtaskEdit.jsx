import React, { Component } from 'react';
import {
  Form, Row, Col, InputGroup, Button,
} from 'react-bootstrap';
import './SubtaskItem.scss';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import accept from '../../assets/accept.svg';
import * as subtaskActions from '../../../actions/subtaskActions';

class SubtaskEdit extends Component {
  handleSubmit = (event) => {
    const {
      _id,
      toggleEditMode,
      subtaskActions: { updateSubtask },
    } = this.props;
    event.preventDefault();
    event.stopPropagation();
    const name = this.input.value.trim().replace(/\s+/g, ' ');
    if (this.form.checkValidity() === false) {
      this.form.classList.add('was-validated');
    } else {
      this.form.classList.remove('was-validated');
      updateSubtask({ payload: name, key: 'name' }, _id);
    }
    toggleEditMode();
  }

  render() {
    const {
      name,
      _id,
    } = this.props;

    return (
      <Row className="subtask-row my-0 mx-0 px-2 rounded">
        <Col className="px-0">
          <Form
            _id={_id}
            className="needs-validation"
            method="POST"
            noValidate
            onSubmit={event => this.handleSubmit(event)}
            ref={(form) => {
              this.form = form;
            }}
          >
            <InputGroup>
              <Form.Control
                className="new-subtask-name was-validated"
                size="sm"
                minLength={1}
                maxLength={40}
                defaultValue={name}
                placeholder="Enter subtask name here..."
                required
                ref={(input) => {
                  this.input = input;
                }}
                onChange={(event) => {
                  if (!event.target.value.trim().length) event.target.value = '';
                }}
              />
              <InputGroup.Append>
                <Button
                  className="subtask-save-button"
                  variant="outline-primary"
                  size="sm"
                  type="submit"
                >
                  <img src={accept} alt="accept" />
                </Button>
              </InputGroup.Append>
              <div className="valid-feedback">Looks good!</div>
              <div className="invalid-feedback">Please provide subtask name.</div>
            </InputGroup>
          </Form>
        </Col>
      </Row>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  subtaskActions: bindActionCreators(subtaskActions, dispatch),
});
export { SubtaskEdit as SubtaskEditComponent };
export default connect(
  null,
  mapDispatchToProps,
)(SubtaskEdit);
