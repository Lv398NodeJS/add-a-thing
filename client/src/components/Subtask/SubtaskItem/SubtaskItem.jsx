import React, { Component } from 'react';
import {
  Form, Row, Col, Dropdown,
} from 'react-bootstrap';
import './SubtaskItem.scss';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as importedSubtaskActions from '../../../actions/subtaskActions';
import threedots from '../../assets/three-dots.svg';

class SubtaskItem extends Component {
  constructor(props) {
    super(props);
    const {
      taskRef,
      subtaskActions: { updateSubtask, deleteSubtask, convertToTask },
    } = this.props;
    this.taskRef = taskRef;
    this.updateSubtask = updateSubtask;
    this.deleteSubtask = deleteSubtask;
    this.convertToTask = convertToTask;
  }

  render() {
    const {
      name,
      completed,
      _id,
      taskStatus,
      toggleEditMode,
    } = this.props;

    return (
      <Row className="subtask-row my-0 mx-0 px-2 rounded">
        <Col
          className="d-flex justify-content-sm-start col-sm-11 px-0 my-1"
          onClick={() => taskStatus !== 'Done' && this.updateSubtask({ payload: !completed, key: 'completed' }, _id)}
        >
          <Form.Check type="checkbox" custom id={_id}>
            <Form.Check.Input
              type="checkbox"
              checked={completed}
              disabled={taskStatus === 'Done'}
              onChange={() => {
                this.updateSubtask({ payload: !completed, key: 'completed' }, _id);
              }}
            />
            <Form.Check.Label>{name}</Form.Check.Label>
          </Form.Check>
        </Col>
        <Col className="d-flex justify-content-sm-end col-sm-1 px-0">
          <Dropdown>
            <Dropdown.Toggle
              variant="light"
              size="sm"
              className="subtask-dropdown-toggle"
              disabled={taskStatus === 'Done'}
            >
              <img src={threedots} alt={threedots} className="threedots-icon" />
            </Dropdown.Toggle>

            <Dropdown.Menu>
              <Dropdown.Item onClick={() => toggleEditMode()}>
                Edit
              </Dropdown.Item>
              <Dropdown.Item onClick={() => this.deleteSubtask(_id)}>
                Delete
              </Dropdown.Item>
              <Dropdown.Item onClick={() => this.convertToTask(_id, { name }, document.URL.split('/').pop())}>
                Convert to Task
              </Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </Col>
      </Row>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  subtaskActions: bindActionCreators(importedSubtaskActions, dispatch),
});
export { SubtaskItem as SubtaskItemComponent };
export default connect(
  null,
  mapDispatchToProps,
)(SubtaskItem);
