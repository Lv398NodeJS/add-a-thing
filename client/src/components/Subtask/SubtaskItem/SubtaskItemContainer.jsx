import React, { Component } from 'react';
import SubtaskItem from './SubtaskItem';

export class SubtaskItemContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditMode: false,
    };
  }

  toggleEditMode = () => {
    this.setState(prevState => ({
      isEditMode: !prevState.isEditMode,
    }));
  }

  render() {
    const {
      id,
      name,
      completed,
      taskStatus,
      taskRef,
    } = this.props;
    const { isEditMode } = this.state;
    return (
      isEditMode
        ? null
        : (
          <SubtaskItem
            id={id}
            name={name}
            completed={completed}
            taskStatus={taskStatus}
            taskRef={taskRef}
          />
        )
    );
  }
}

export default SubtaskItemContainer;
