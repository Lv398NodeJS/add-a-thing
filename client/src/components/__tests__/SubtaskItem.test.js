import React from 'react';
import { shallow } from 'enzyme';
import { SubtaskItemComponent } from '../Subtask/SubtaskItem/SubtaskItem';

describe('SubtaskItem component', () => {
  const subtaskActions = {
    changeSubtaskStatus: jest.fn(),
    deleteSubtask: jest.fn(),
    convertToTask: jest.fn(),
  };
  it('should render correctly and match the snapshot', () => {
    const subtaskItemComponent = shallow(
      <SubtaskItemComponent subtaskActions={subtaskActions} />,
    );
    expect(subtaskItemComponent).toMatchSnapshot();
  });

  // it('should render subtask text properly', () => {
  //   const text = 'Subtask test text';
  //   const subtaskItemComponent = shallow(
  //     <SubtaskItemComponent
  //       text={text}
  //       subtaskActions={subtaskActions}
  //     />,
  //   );
  //   expect(subtaskItemComponent.find('FormCheckLabel').text()).toBe(text);
  // });

  it('should render subtask checkbox properly according to subtask completion status', () => {
    const completed = true;
    const subtaskItemComponent = shallow(
      <SubtaskItemComponent
        completed={completed}
        subtaskActions={subtaskActions}
      />,
    );
    expect(subtaskItemComponent.find('FormCheckInput').props().checked).toBe(completed);
  });
});
