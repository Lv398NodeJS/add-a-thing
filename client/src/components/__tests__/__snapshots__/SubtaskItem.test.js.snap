// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubtaskItem component should render correctly and match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SubtaskItem
    subtaskActions={
      Object {
        "changeSubtaskStatus": [MockFunction],
        "convertToTask": [MockFunction],
        "deleteSubtask": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ForwardRef
          className="d-flex justify-content-sm-start col-sm-11 px-0 my-1"
          onClick={[Function]}
        >
          <ForwardRef
            custom={true}
            disabled={false}
            inline={false}
            isInvalid={false}
            isValid={false}
            title=""
            type="checkbox"
          >
            <ForwardRef
              disabled={false}
              onChange={[Function]}
              type="checkbox"
            />
            <ForwardRef
              type="checkbox"
            />
          </ForwardRef>
        </ForwardRef>,
        <ForwardRef
          className="d-flex justify-content-sm-end col-sm-1 px-0"
        >
          <ForwardRef
            navbar={false}
          >
            <ForwardRef(render)
              className="subtask-dropdown-toggle"
              disabled={false}
              size="sm"
              variant="light"
            >
              <img
                alt="Menu"
                className="threedots-icon"
                src="three-dots.svg"
              />
            </ForwardRef(render)>
            <ForwardRef
              alignRight={false}
              flip={true}
            >
              <ForwardRef
                as={[Function]}
                disabled={false}
                onClick={[Function]}
              >
                Edit
              </ForwardRef>
              <ForwardRef
                as={[Function]}
                disabled={false}
                onClick={[Function]}
              >
                Delete
              </ForwardRef>
              <ForwardRef
                as={[Function]}
                disabled={false}
                onClick={[Function]}
              >
                Convert to Task
              </ForwardRef>
            </ForwardRef>
          </ForwardRef>
        </ForwardRef>,
      ],
      "className": "subtask-row my-0 mx-0 px-2 rounded",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ForwardRef
            custom={true}
            disabled={false}
            inline={false}
            isInvalid={false}
            isValid={false}
            title=""
            type="checkbox"
          >
            <ForwardRef
              disabled={false}
              onChange={[Function]}
              type="checkbox"
            />
            <ForwardRef
              type="checkbox"
            />
          </ForwardRef>,
          "className": "d-flex justify-content-sm-start col-sm-11 px-0 my-1",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <ForwardRef
                disabled={false}
                onChange={[Function]}
                type="checkbox"
              />,
              <ForwardRef
                type="checkbox"
              />,
            ],
            "custom": true,
            "disabled": false,
            "id": undefined,
            "inline": false,
            "isInvalid": false,
            "isValid": false,
            "title": "",
            "type": "checkbox",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "checked": undefined,
                "disabled": false,
                "onChange": [Function],
                "type": "checkbox",
              },
              "ref": null,
              "rendered": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "type": "checkbox",
                },
                "displayName": "FormCheckInput",
                "render": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": undefined,
                "type": "checkbox",
              },
              "ref": null,
              "rendered": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "type": "checkbox",
                },
                "displayName": "FormCheckLabel",
                "render": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Input": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "type": "checkbox",
              },
              "displayName": "FormCheckInput",
              "render": [Function],
            },
            "Label": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "type": "checkbox",
              },
              "displayName": "FormCheckLabel",
              "render": [Function],
            },
            "defaultProps": Object {
              "disabled": false,
              "inline": false,
              "isInvalid": false,
              "isValid": false,
              "title": "",
              "type": "checkbox",
            },
            "displayName": "FormCheck",
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Col",
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ForwardRef
            navbar={false}
          >
            <ForwardRef(render)
              className="subtask-dropdown-toggle"
              disabled={false}
              size="sm"
              variant="light"
            >
              <img
                alt="Menu"
                className="threedots-icon"
                src="three-dots.svg"
              />
            </ForwardRef(render)>
            <ForwardRef
              alignRight={false}
              flip={true}
            >
              <ForwardRef
                as={[Function]}
                disabled={false}
                onClick={[Function]}
              >
                Edit
              </ForwardRef>
              <ForwardRef
                as={[Function]}
                disabled={false}
                onClick={[Function]}
              >
                Delete
              </ForwardRef>
              <ForwardRef
                as={[Function]}
                disabled={false}
                onClick={[Function]}
              >
                Convert to Task
              </ForwardRef>
            </ForwardRef>
          </ForwardRef>,
          "className": "d-flex justify-content-sm-end col-sm-1 px-0",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <ForwardRef(render)
                className="subtask-dropdown-toggle"
                disabled={false}
                size="sm"
                variant="light"
              >
                <img
                  alt="Menu"
                  className="threedots-icon"
                  src="three-dots.svg"
                />
              </ForwardRef(render)>,
              <ForwardRef
                alignRight={false}
                flip={true}
              >
                <ForwardRef
                  as={[Function]}
                  disabled={false}
                  onClick={[Function]}
                >
                  Edit
                </ForwardRef>
                <ForwardRef
                  as={[Function]}
                  disabled={false}
                  onClick={[Function]}
                >
                  Delete
                </ForwardRef>
                <ForwardRef
                  as={[Function]}
                  disabled={false}
                  onClick={[Function]}
                >
                  Convert to Task
                </ForwardRef>
              </ForwardRef>,
            ],
            "navbar": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <img
                  alt="Menu"
                  className="threedots-icon"
                  src="three-dots.svg"
                />,
                "className": "subtask-dropdown-toggle",
                "disabled": false,
                "size": "sm",
                "variant": "light",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "Menu",
                  "className": "threedots-icon",
                  "src": "three-dots.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "displayName": "Bootstrap(DropdownToggle)",
                "propTypes": undefined,
                "render": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "alignRight": false,
                "children": Array [
                  <ForwardRef
                    as={[Function]}
                    disabled={false}
                    onClick={[Function]}
                  >
                    Edit
                  </ForwardRef>,
                  <ForwardRef
                    as={[Function]}
                    disabled={false}
                    onClick={[Function]}
                  >
                    Delete
                  </ForwardRef>,
                  <ForwardRef
                    as={[Function]}
                    disabled={false}
                    onClick={[Function]}
                  >
                    Convert to Task
                  </ForwardRef>,
                ],
                "flip": true,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "as": [Function],
                    "children": "Edit",
                    "disabled": false,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Edit",
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "defaultProps": Object {
                      "as": [Function],
                      "disabled": false,
                    },
                    "displayName": "DropdownItem",
                    "render": [Function],
                  },
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "as": [Function],
                    "children": "Delete",
                    "disabled": false,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Delete",
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "defaultProps": Object {
                      "as": [Function],
                      "disabled": false,
                    },
                    "displayName": "DropdownItem",
                    "render": [Function],
                  },
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "as": [Function],
                    "children": "Convert to Task",
                    "disabled": false,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Convert to Task",
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "defaultProps": Object {
                      "as": [Function],
                      "disabled": false,
                    },
                    "displayName": "DropdownItem",
                    "render": [Function],
                  },
                },
              ],
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "alignRight": false,
                  "flip": true,
                },
                "displayName": "DropdownMenu",
                "render": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Divider": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "role": "separator",
              },
              "displayName": "DropdownDivider",
              "render": [Function],
            },
            "Header": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "role": "heading",
              },
              "displayName": "DropdownHeader",
              "render": [Function],
            },
            "Item": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "as": [Function],
                "disabled": false,
              },
              "displayName": "DropdownItem",
              "render": [Function],
            },
            "Menu": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "alignRight": false,
                "flip": true,
              },
              "displayName": "DropdownMenu",
              "render": [Function],
            },
            "Toggle": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "displayName": "Bootstrap(DropdownToggle)",
              "propTypes": undefined,
              "render": [Function],
            },
            "defaultProps": Object {
              "navbar": false,
            },
            "displayName": "Dropdown",
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Col",
          "render": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": undefined,
      "displayName": "Bootstrap(Row)",
      "propTypes": undefined,
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef
            className="d-flex justify-content-sm-start col-sm-11 px-0 my-1"
            onClick={[Function]}
          >
            <ForwardRef
              custom={true}
              disabled={false}
              inline={false}
              isInvalid={false}
              isValid={false}
              title=""
              type="checkbox"
            >
              <ForwardRef
                disabled={false}
                onChange={[Function]}
                type="checkbox"
              />
              <ForwardRef
                type="checkbox"
              />
            </ForwardRef>
          </ForwardRef>,
          <ForwardRef
            className="d-flex justify-content-sm-end col-sm-1 px-0"
          >
            <ForwardRef
              navbar={false}
            >
              <ForwardRef(render)
                className="subtask-dropdown-toggle"
                disabled={false}
                size="sm"
                variant="light"
              >
                <img
                  alt="Menu"
                  className="threedots-icon"
                  src="three-dots.svg"
                />
              </ForwardRef(render)>
              <ForwardRef
                alignRight={false}
                flip={true}
              >
                <ForwardRef
                  as={[Function]}
                  disabled={false}
                  onClick={[Function]}
                >
                  Edit
                </ForwardRef>
                <ForwardRef
                  as={[Function]}
                  disabled={false}
                  onClick={[Function]}
                >
                  Delete
                </ForwardRef>
                <ForwardRef
                  as={[Function]}
                  disabled={false}
                  onClick={[Function]}
                >
                  Convert to Task
                </ForwardRef>
              </ForwardRef>
            </ForwardRef>
          </ForwardRef>,
        ],
        "className": "subtask-row my-0 mx-0 px-2 rounded",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ForwardRef
              custom={true}
              disabled={false}
              inline={false}
              isInvalid={false}
              isValid={false}
              title=""
              type="checkbox"
            >
              <ForwardRef
                disabled={false}
                onChange={[Function]}
                type="checkbox"
              />
              <ForwardRef
                type="checkbox"
              />
            </ForwardRef>,
            "className": "d-flex justify-content-sm-start col-sm-11 px-0 my-1",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <ForwardRef
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />,
                <ForwardRef
                  type="checkbox"
                />,
              ],
              "custom": true,
              "disabled": false,
              "id": undefined,
              "inline": false,
              "isInvalid": false,
              "isValid": false,
              "title": "",
              "type": "checkbox",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "checked": undefined,
                  "disabled": false,
                  "onChange": [Function],
                  "type": "checkbox",
                },
                "ref": null,
                "rendered": null,
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": Object {
                    "type": "checkbox",
                  },
                  "displayName": "FormCheckInput",
                  "render": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": undefined,
                  "type": "checkbox",
                },
                "ref": null,
                "rendered": null,
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": Object {
                    "type": "checkbox",
                  },
                  "displayName": "FormCheckLabel",
                  "render": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "Input": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "type": "checkbox",
                },
                "displayName": "FormCheckInput",
                "render": [Function],
              },
              "Label": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "type": "checkbox",
                },
                "displayName": "FormCheckLabel",
                "render": [Function],
              },
              "defaultProps": Object {
                "disabled": false,
                "inline": false,
                "isInvalid": false,
                "isValid": false,
                "title": "",
                "type": "checkbox",
              },
              "displayName": "FormCheck",
              "render": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ForwardRef
              navbar={false}
            >
              <ForwardRef(render)
                className="subtask-dropdown-toggle"
                disabled={false}
                size="sm"
                variant="light"
              >
                <img
                  alt="Menu"
                  className="threedots-icon"
                  src="three-dots.svg"
                />
              </ForwardRef(render)>
              <ForwardRef
                alignRight={false}
                flip={true}
              >
                <ForwardRef
                  as={[Function]}
                  disabled={false}
                  onClick={[Function]}
                >
                  Edit
                </ForwardRef>
                <ForwardRef
                  as={[Function]}
                  disabled={false}
                  onClick={[Function]}
                >
                  Delete
                </ForwardRef>
                <ForwardRef
                  as={[Function]}
                  disabled={false}
                  onClick={[Function]}
                >
                  Convert to Task
                </ForwardRef>
              </ForwardRef>
            </ForwardRef>,
            "className": "d-flex justify-content-sm-end col-sm-1 px-0",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <ForwardRef(render)
                  className="subtask-dropdown-toggle"
                  disabled={false}
                  size="sm"
                  variant="light"
                >
                  <img
                    alt="Menu"
                    className="threedots-icon"
                    src="three-dots.svg"
                  />
                </ForwardRef(render)>,
                <ForwardRef
                  alignRight={false}
                  flip={true}
                >
                  <ForwardRef
                    as={[Function]}
                    disabled={false}
                    onClick={[Function]}
                  >
                    Edit
                  </ForwardRef>
                  <ForwardRef
                    as={[Function]}
                    disabled={false}
                    onClick={[Function]}
                  >
                    Delete
                  </ForwardRef>
                  <ForwardRef
                    as={[Function]}
                    disabled={false}
                    onClick={[Function]}
                  >
                    Convert to Task
                  </ForwardRef>
                </ForwardRef>,
              ],
              "navbar": false,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <img
                    alt="Menu"
                    className="threedots-icon"
                    src="three-dots.svg"
                  />,
                  "className": "subtask-dropdown-toggle",
                  "disabled": false,
                  "size": "sm",
                  "variant": "light",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "Menu",
                    "className": "threedots-icon",
                    "src": "three-dots.svg",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": undefined,
                  "displayName": "Bootstrap(DropdownToggle)",
                  "propTypes": undefined,
                  "render": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "alignRight": false,
                  "children": Array [
                    <ForwardRef
                      as={[Function]}
                      disabled={false}
                      onClick={[Function]}
                    >
                      Edit
                    </ForwardRef>,
                    <ForwardRef
                      as={[Function]}
                      disabled={false}
                      onClick={[Function]}
                    >
                      Delete
                    </ForwardRef>,
                    <ForwardRef
                      as={[Function]}
                      disabled={false}
                      onClick={[Function]}
                    >
                      Convert to Task
                    </ForwardRef>,
                  ],
                  "flip": true,
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "as": [Function],
                      "children": "Edit",
                      "disabled": false,
                      "onClick": [Function],
                    },
                    "ref": null,
                    "rendered": "Edit",
                    "type": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "defaultProps": Object {
                        "as": [Function],
                        "disabled": false,
                      },
                      "displayName": "DropdownItem",
                      "render": [Function],
                    },
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "as": [Function],
                      "children": "Delete",
                      "disabled": false,
                      "onClick": [Function],
                    },
                    "ref": null,
                    "rendered": "Delete",
                    "type": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "defaultProps": Object {
                        "as": [Function],
                        "disabled": false,
                      },
                      "displayName": "DropdownItem",
                      "render": [Function],
                    },
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "as": [Function],
                      "children": "Convert to Task",
                      "disabled": false,
                      "onClick": [Function],
                    },
                    "ref": null,
                    "rendered": "Convert to Task",
                    "type": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "defaultProps": Object {
                        "as": [Function],
                        "disabled": false,
                      },
                      "displayName": "DropdownItem",
                      "render": [Function],
                    },
                  },
                ],
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": Object {
                    "alignRight": false,
                    "flip": true,
                  },
                  "displayName": "DropdownMenu",
                  "render": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "Divider": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "role": "separator",
                },
                "displayName": "DropdownDivider",
                "render": [Function],
              },
              "Header": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "role": "heading",
                },
                "displayName": "DropdownHeader",
                "render": [Function],
              },
              "Item": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "as": [Function],
                  "disabled": false,
                },
                "displayName": "DropdownItem",
                "render": [Function],
              },
              "Menu": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "alignRight": false,
                  "flip": true,
                },
                "displayName": "DropdownMenu",
                "render": [Function],
              },
              "Toggle": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "displayName": "Bootstrap(DropdownToggle)",
                "propTypes": undefined,
                "render": [Function],
              },
              "defaultProps": Object {
                "navbar": false,
              },
              "displayName": "Dropdown",
              "render": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "Col",
            "render": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "displayName": "Bootstrap(Row)",
        "propTypes": undefined,
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
